### PR TITLE
[NEEDS REVIEW] Throw/Pick Up Drones/Various Animals! (Pick up /living/ framework + Throw mob_holders)

### PR DESCRIPTION
**Pull request by Ziiro**
:cl: ChuckTheSheep, FlattestGuitar, Ziiro
add: You can now pick up (and wear!) certain simple mobs/pais/monkeys with alt+click
/:cl:

This PR is mostly made up of https://github.com/tgstation/tgstation/pull/35532 but it also fixes worn mob breathing/atmos (thanks FlattestGuitar) and makes some adjustments for futureproofing so that carbons could be held if that's a wanted feature.

### CODE REVIEW

**Comment by MrDoomBringer** [comment]
does this reactor climbing on people's backs?

**Comment by Ziiro** [comment]
No

File: code/modules/mob/living/inhand_holder.dm [hunk 1]
```diff
@@ -5,80 +5,140 @@
...
+
+	if(M)
+		M.setDir(SOUTH)
+		held_mob = M
+		M.forceMove(src)
+		appearance = M.appearance
+		name = M.name
+		desc = M.desc
+
+	if(_can_head_override)
```
Comment (outbools) [inline code/modules/mob/living/inhand_holder.dm]: 🤔

**Comment by Wilchenx** [comment]
@Ziiro https://github.com/tgstation/tgstation/pull/25114

**Comment by Ziiro** [comment]
Man going around that would be a fucking mess. I wonder if they should be a neck item instead.

**Comment by ExcessiveUseOfCobblestone** [comment]
@Ziiro it's `dynamic_hair_suffix` for the first problem

**Comment by Ziiro** [comment]
This is ready for judgement

**Comment by Tlaltecuhtli** [comment]
assault pocket monkey time

**Comment by Ziiro** [comment]
they're a bulky item, so the only way you can store them is in a BOH. Also they can't breathe in a BOH. That being said, when you throw a mob holder it counts as *throwing a mob*. So you could just have a backpack of holding full of monkeys you throw at people to knock them down.

**Comment by davethwave** [comment]
What about other players? Can you just stuff people or corpses in Boh's?

**Comment by Ziiro** [comment]
No. In this version* of this PR, there is no way to pick up anything other than Monkeys, Station Pets, Drones, and pAIs.

*In the Oracle version, I made getting dwarfed in genetics pick up-able. It's total hilarious/fun, but I feel like there's some balance repercussions that would arise out of that on a more robust playerbase that doesn't occur on Oracle.

File: code/game/machinery/suit_storage_unit.dm [hunk 2]
```diff
@@ -305,6 +305,9 @@
 
 /obj/machinery/suit_storage_unit/attackby(obj/item/I, mob/user, params)
 	if(state_open && is_operational())
+		if(istype(I, /obj/item/clothing/head/mob_holder))
+			to_chat(user, "<span class='warning'>That doesn't fit in the unit like that! Try another method!</span>")
```
Comment (Fox-McCloud) [inline code/game/machinery/suit_storage_unit.dm]: This is gross, and snowflakey. This should be expanded to be something that is either a flag or a var on an object that, if set, prevents it from being placed inside this.

File: code/game/machinery/washing_machine.dm [hunk 3]
```diff
@@ -213,6 +213,10 @@
 		update_icon()
 		return
 
+	if(istype(W, /obj/item/clothing/head/mob_holder))
```
Comment (Fox-McCloud) [inline code/game/machinery/washing_machine.dm]: same here.

File: code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm [hunk 4]
```diff
@@ -90,6 +90,9 @@ God bless America.
 	if(istype(I, /obj/item/reagent_containers/food/snacks/deepfryholder))
 		to_chat(user, "<span class='userdanger'>Your cooking skills are not up to the legendary Doublefry technique.</span>")
 		return
+	if(istype(I,/obj/item/clothing/head/mob_holder))
```
Comment (Fox-McCloud) [inline code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm]: and here.

File: code/modules/mob/living/carbon/carbon.dm [hunk 5]
```diff
@@ -150,19 +150,26 @@
...
-				stop_pulling()
-				if(has_trait(TRAIT_PACIFISM))
-					to_chat(src, "<span class='notice'>You gently let go of [throwable_mob].</span>")
-				var/turf/start_T = get_turf(loc) //Get the start and target tile for the descriptors
-				var/turf/end_T = get_turf(target)
-				if(start_T && end_T)
-					add_logs(src, throwable_mob, "thrown", addition="grab from tile in [AREACOORD(start_T)] towards tile at [AREACOORD(end_T)]")
+	var/mob/living/throwable_mob
+
+	if(istype(I, /obj/item/clothing/head/mob_holder))
```
Comment (Fox-McCloud) [inline code/modules/mob/living/carbon/carbon.dm]: This is also pretty gross; snowflaking behavior out is a blatant violation of OOP.

File: code/modules/mob/living/inhand_holder.dm [hunk 6]
```diff
@@ -5,80 +5,141 @@
...
+	var/obj/item/clothing/head/mob_holder/holder = new(get_turf(src), src, (istext(can_be_held) ? can_be_held : ""), 'icons/mob/animals_held.dmi', 'icons/mob/animals_held_lh.dmi', 'icons/mob/animals_held_rh.dmi')
+	holder.w_class = 1
+	return holder
+
+/mob/living/simple_animal/hostile/lizard/generate_mob_holder()
+	var/obj/item/clothing/head/mob_holder/holder = new(get_turf(src), src, "lizard", 'icons/mob/animals_held.dmi', 'icons/mob/animals_held_lh.dmi', 'icons/mob/animals_held_rh.dmi', TRUE)
+	return holder
+
+/mob/living/proc/mob_try_pickup(mob/living/user)
+	if(!ishuman(user) || !src.Adjacent(user) || user.incapacitated() || !can_be_held)
```
Comment (Fox-McCloud) [inline code/modules/mob/living/inhand_holder.dm]: no implied `src`. Here or anywhere else unless you ever change `src` or have poorly named variables.

File: code/modules/recycling/disposal/bin.dm [hunk 7]
```diff
@@ -105,8 +105,17 @@
 		return ..()
 
 /obj/machinery/disposal/proc/place_item_in_disposal(obj/item/I, mob/user)
-	I.forceMove(src)
-	user.visible_message("[user.name] places \the [I] into \the [src].", "<span class='notice'>You place \the [I] into \the [src].</span>")
+	if(istype(I, /obj/item/clothing/head/mob_holder))
```
Comment (Fox-McCloud) [inline code/modules/recycling/disposal/bin.dm]: ❄️

File: code/modules/research/destructive_analyzer.dm [hunk 8]
```diff
@@ -152,3 +152,8 @@ Note: Must be placed within 3 tiles of the R&D Console
 	loaded_item = null
 	update_icon()
 	return TRUE
+
+/obj/machinery/rnd/destructive_analyzer/attackby(obj/item/I, mob/user)
+	if(istype(I, /obj/item/clothing/head/mob_holder))
+		to_chat(user, "<span class='warning'>The [src.name] smartly refuses \the [I].</span>")
```
Comment (Fox-McCloud) [inline code/modules/research/destructive_analyzer.dm]: no `\the`; BYOND handles this already. also no implied `src` either.

**Comment by Fox-McCloud** [comment]
I kinda wonder if mob holders should be a clothing subtype at all. As far as I know, something doens't necessarily have to be a clothing/head subtype to be worn on the head and for it to be rendered properly.

As far as I know, this can all be controlled with clothing flags.

The fact that so many clothing exceptions have to be made inclines me to believe that them being a clothing subtype may not be the best thing.

**Comment by Kagus** [comment]
Paper can be worn on the head, and the chaplain's atheist fedora can also be worn on the head, despite not actually being a hat.

**Comment by Doppeltongue** [comment]
If manlets don't want to get tossed into the SM, they should get some mutadone.

**Comment by Ziiro** [comment]
@Fox-McCloud I could switch it over to a basic obj type? I think the warning cone is probably a good example of what I have in mind.

also with regards to ❄️: I could make it an item trait, but I'm not sure what all of those things have in common I could put under one sort of flag (`NO_DESTROY`?). I could make it a set of flags with each device having a check: i.e, `NO_FRY`, `NO_WASH`, `NO_ANALYZE`, `NO_SUIT_STORAGE`, but I'm not actually sure that's any better. Is it?

**Comment by ExcessiveUseOfCobblestone** [comment]
Don't think it should be a component, a living flag

**Comment by Ziiro** [comment]
It's already a living var. kevinz000 put that in a different animal pickup PR

https://github.com/tgstation/tgstation/blob/b9c5bad6055496cc38dba541902b55ae1c50234d/code/modules/mob/living/living_defines.dm#L106

I'll fix the ❄️ I just need to know what actually works. A singular item flag, or multiple?

I figure I can re-organize it under `obj/mob_holder` and still make it able to be worn (Thanks for the Atheist Fedora blueprint on that)

**Comment by Ziiro** [comment]
why the fuck didn't anyone just tell me to use abstract, christ jesus it's that easy (for most of the problems)

**Comment by ExcessiveUseOfCobblestone** [comment]
Not everyone can be an epic silicon like your avatar, anon

**Comment by TheDreamweaver** [comment]
Can we get a review up in here.

**Comment by GrayRachnid** [comment]
alt+click for this seems very hacky.

File: code/modules/mob/living/inhand_holder.dm [hunk 9]
```diff
@@ -1,84 +1,145 @@
...
+		return FALSE
+	if(src == user)
+		to_chat(user, "<span class='warning'>You can't pick yourself up.</span>")
+		return FALSE
+	if(user.can_be_held)
+		to_chat(user, "<span class='warning'>You're too small to pick that up.</span>")
+		return FALSE
+	visible_message("<span class='warning'>[user] starts picking up [src].</span>", \
+					"<span class='userdanger'>[user] starts picking you up!</span>")
+	if(!do_after(user, 20, target = src))
```
Comment (Ziiro) [inline code/modules/mob/living/inhand_holder.dm]: Shit I forgot I can use `2 SECONDS` here, neat.

**Comment by nicbn** [comment]
alt+click sucks
As ctrl+click pull already does the job for grab intent, make grab intent pick the mob up (if it can't be picked up, it behaves like pull)

**Comment by Ziiro** [comment]
Problem is I'd like to have this framework available for carbons as well. If I tie it to grab intent, that will definitely suck.

**Comment by ExcessiveUseOfCobblestone** [comment]
whats the issue with altclick again?

**Comment by Kagus** [comment]
> whats the issue with altclick again?

We don't approve of alt culture around here.

**Comment by nicbn** [comment]
It's not intuitive and it is already used for a lot of unrelated things

**Comment by ExcessiveUseOfCobblestone** [comment]
picking up is the _alternative_ to furthering the grab into choke or whatever so I don't find that not intuitive. This is probably one of the few things that DOES make sense to toss on alt.

Having normal grab function conflict with this for the sole sake of appeasing non-alt users is ???. At the very least, it should be a singular control so we aren't further taking up click possibilities.

It's nice to know this PR is so good the only thing left to bikeshed is people's preferences on what to click though.

**Comment by Kagus** [comment]
Hey, nobody can resist a good game of dwarf tossing. Not even the dwarfs!

Especially not the dwarfs.

**Comment by Ziiro** [comment]
Going on hiatus from SS13 again, if someone else wants to bother with this they have my support.

**Pull request closed without merging**